src/main/java/org/opensearch/neuralsearch/query/HybridQuery.java

```diff
@@ -106,22 +111,40 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
             return new MatchNoDocsQuery("empty HybridQuery");
         }
 
-        boolean actuallyRewritten = false;
-        List<Query> rewrittenSubQueries = new ArrayList<>();
+        HybridQueryRewriteCollectorManager manager = new HybridQueryRewriteCollectorManager(indexSearcher);
+        List<Callable<Void>> queryRewriteTasks = new ArrayList<>();
+        List<HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>>> collectors = new ArrayList<>();
         for (Query subQuery : subQueries) {
-            Query rewrittenSub = subQuery.rewrite(indexSearcher);
-            /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
-               queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
-               perform better. For hybrid query we need to track progress of re-write for all sub-queries */
-            actuallyRewritten |= rewrittenSub != subQuery;
-            rewrittenSubQueries.add(rewrittenSub);
+            HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector = manager.newCollector();
+            collectors.add(collector);
+            queryRewriteTasks.add(() -> rewriteQuery(subQuery, collector));
         }
 
-        if (actuallyRewritten) {
-            return new HybridQuery(rewrittenSubQueries);
+        HybridQueryExecutor.getExecutor().invokeAll(queryRewriteTasks);
+
+        final boolean anyQueryRewrite = manager.anyQueryRewrite(collectors);
+        if (anyQueryRewrite == false) {
+            return super.rewrite(indexSearcher);
         }
+        final List<Query> rewrittenSubQueries = manager.getRewriteQueries(collectors);
```

Comment:
how we are ensuring that all the queries are rewritten?

Reply:
if query rewrite is not done, you will have Query as null object in the list.

Reply:
The below command make sure that all tasks are completed `HybridQueryExecutor.getExecutor().invokeAll(queryRewriteTasks);`